### PR TITLE
automatic function

### PR DESCRIPTION
**Pull request by pmcurtin**
#47

### CODE REVIEW

**Comment by pmcurtin** [comment]
@apoorvkh Was this kind of what you were thinking of? The `automatic` function returns the `hostnames` and `workers_per_host` arguments for `launch`... The user would have to call `hosts, workers = automatic()` and then manually pass the arguments to `launch`. We could just add a flag to `launch` that causes it to call `automatic` internally and use its output rather than the hostnames and workers that the user passes. 

We can also rename `automatic` to something more descriptive like ` automatic_topology` (idk).

File: src/torchrunx/utils.py [hunk 1]
```diff
@@ -21,6 +25,23 @@ def get_open_port() -> int:
...
+    :return: Allocation hosts and workers per host
+    :rtype: tuple[list[str], int]
+    """
+
+    if "SLURM_JOB_ID" not in os.environ:
+        _cpus = os.cpu_count()
+        cpus = 1 if _cpus is None else _cpus
+        gpus = torch.cuda.device_count()
+        return (["localhost"], cpus if gpus == 0 else gpus)
+
```
Comment (apoorvkh) [inline src/torchrunx/utils.py:41]: else:
    return slurm_hosts(), slurm_workers()

File: src/torchrunx/utils.py [hunk 2]
```diff
@@ -21,6 +25,23 @@ def get_open_port() -> int:
...
 
+def automatic() -> tuple[list[str], int]:
+    """
+    Automatically determine allocation sizes
+
+    :return: Allocation hosts and workers per host
+    :rtype: tuple[list[str], int]
+    """
+
+    if "SLURM_JOB_ID" not in os.environ:
```
Comment (apoorvkh) [inline src/torchrunx/utils.py:36]: let's actually make a

```python
def in_slurm_job() -> bool:
    return "SLURM_JOB_ID" in os.environ
```

so we can more consistently call that in various places.

File: src/torchrunx/utils.py [hunk 3]
```diff
@@ -21,6 +25,23 @@ def get_open_port() -> int:
     return port
 
 
+def automatic() -> tuple[list[str], int]:
```
Comment (apoorvkh) [inline src/torchrunx/utils.py:28]: Maybe split this into `auto_hosts` and `auto_workers`?

Comment (apoorvkh) [reply src/torchrunx/utils.py:28]: additionally, let's put it in `auto.py`

Comment (apoorvkh) [reply src/torchrunx/utils.py:28]: or a more general `environment.py` and then we put all the `auto` and `slurm` functions in there?

**Comment by apoorvkh** [comment]
Yeah, close to what I was suggesting! See comments.

And I agree, we should add flags to launch: e.g. `launch(..., auto=True, hostnames=None, workers=None)` where this would infer `if hostnames is None: hostnames = auto_hosts()` etc.

File: src/torchrunx/environment.py [hunk 4]
```diff
@@ -29,10 +35,40 @@ def slurm_workers() -> int:
...
+    """
+    Automatically determine hostname list
+
+    :return: Hostnames in Slurm allocation, or ['localhost']
+    :rtype: list[str]
+    """
+    if in_slurm_job():
+        return ["localhost"]
+
+    return slurm_hosts()
```
Comment (apoorvkh) [inline src/torchrunx/environment.py:57]: Aha, I think the logic is switched :)
also for `auto_workers`

File: src/torchrunx/environment.py [hunk 5]
```diff
@@ -29,10 +35,40 @@ def slurm_workers() -> int:
...
+
+    :return: Workers per host
+    :rtype: int
+    """
+
+    if in_slurm_job:
+        _cpus = os.cpu_count()
+        cpus = 1 if _cpus is None else _cpus
+        gpus = torch.cuda.device_count()
+        return cpus if gpus == 0 else gpus
```
Comment (apoorvkh) [inline src/torchrunx/environment.py:72]: I think you could simply write this as `return torch.cuda.device_count() or os.cpu_count() or 1`

File: src/torchrunx/launcher.py [hunk 6]
```diff
@@ -97,6 +98,7 @@ class Launcher:
     )
     env_file: str | os.PathLike | None = None
     timeout: int = 600
+    auto: bool = False
```
Comment (apoorvkh) [inline src/torchrunx/launcher.py:101]: I would consider change the order and defaults to
```python
auto: bool = False
hostnames: list[str] | None =["localhost"]
workers_per_host: int | list[int] | None = 1
```

Then, if the user wants `auto` mode: they would do `Launcher(auto=True, hostnames=None, workers_per_host=None, ...)`

**Review by apoorvkh — APPROVED** [review]
LGTM -- feel free to merge once tests pass

**Pull request merged**